radix-engine-interface/src/time/date_time.rs

```diff
@@ -0,0 +1,512 @@
+
+const UNIX_EPOCH_YEAR: u32 = 1970;
+const SECONDS_IN_A_NON_LEAP_YEAR: u64 = 365 * 24 * 60 * 60;
+const SECONDS_IN_A_LEAP_YEAR: u64 = 366 * 24 * 60 * 60;
+const DAYS_PER_4Y: i64 = 365 * 4 + 1;
+const DAYS_PER_100Y: i64 = 365 * 100 + 24;
+const DAYS_PER_400Y: i64 = 365 * 400 + 97;
+const LEAP_YEAR_DAYS_IN_MONTHS: [u8; 12] = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
+// A shift (in seconds) from the Unix epoch (1970-01-01 00:00:00)
+// to a base date that is a multiple of a 400-year leap cycle.
+// Used in Instant -> DateTime conversion.
+// The date we're using is 2000-03-01 00:00:00, for two reasons:
+// a) It's a multiple of 400, to make it easier to work with leap years
+// b) We're also shifting the month to 1st March, so that
+//    the extra day on leap years is added to the last month (Feb),
+//    not in the middle of a year (makes some calculations easier)
+const SHIFT_FROM_UNIX_TIME_TO_MARCH_Y2K: i64 = 946684800 + 86400 * (31 + 29);
+const MIN_SUPPORTED_TIMESTAMP: i64 = -62135596800;
+const MAX_SUPPORTED_TIMESTAMP: i64 = 135536014634284799;
+
+#[derive(Encode, Decode, TypeId, PartialEq, Eq, Copy, Clone, Debug)]
+pub struct DateTime {
+    year: u32,
+    month: u8,
+    day_of_month: u8,
+    hour: u8,
+    minute: u8,
+    second: u8,
+}
+
+impl DateTime {
+    pub fn new(
+        year: u32,
+        month: u8,
+        day_of_month: u8,
+        hour: u8,
+        minute: u8,
+        second: u8,
+    ) -> Result<Self, String> {
+        if year <= 0 {
+            return Err("Invalid year. Expected a value strictly greater than 0".to_string());
+        }
+
+        if month < 1 || month > 12 {
+            return Err(
+                "Invalid month. Expected a value between 1 (inclusive) and 12 (inclusive)"
+                    .to_string(),
+            );
+        }
+
+        if day_of_month < 1 ||
+            // Check leap year Feb + all other months
+            day_of_month > LEAP_YEAR_DAYS_IN_MONTHS[(month - 1) as usize] ||
+            // Check Feb on non-leap years
+            (!Self::is_leap_year(year) && month == 2 && day_of_month > 28)
+        {
+            return Err("Invalid day of month.\
+                Expected a value between 1 (inclusive)\
+                and, depending on a month, 28, 29 (Feb on a leap year), 30 or 31 (inclusive)"
+                .to_string());
+        }
+
+        if hour > 23 {
+            return Err(
+                "Invalid hour. Expected a value between 0 (inclusive) and 23 (inclusive)"
+                    .to_string(),
+            );
+        }
+
+        if minute > 59 {
+            return Err(
+                "Invalid minute. Expected a value between 0 (inclusive) and 59 (inclusive)"
+                    .to_string(),
+            );
+        }
+
+        if second > 59 {
+            return Err(
+                "Invalid second. Expected a value between 0 (inclusive) and 59 (inclusive)"
+                    .to_string(),
```

Comment:
Regarding inclusiveness and exclusiveness, are we following a standard?

Reply:
I believe 0-23 (0-59 respectively) is more or less a standard

Reply:
Maybe we can document the range of  `year`, `month`, `day_of_month`, `hour`, `minute`, `second`  on the constructor `new` if no standard has been strictly followed.

Reply:
Hmh, I'd prefer not to reference the standard in the doc. For example, apparently some versions of it allow hour=24 (https://en.wikipedia.org/wiki/ISO_8601#:~:text=As%20of%20ISO%208601%2D1%3A2019/Amd,permitted%20in%20earlier%20versions%20of%20the%20standard.) and `60` seems to be a valid value for `second` (for leap seconds, which we don't support).